src/expr/src/expr/expr_array_distinct.rs

```diff
@@ -0,0 +1,255 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use itertools::Itertools;
+use risingwave_common::array::*;
+use risingwave_common::row::OwnedRow;
+use risingwave_common::types::{DataType, Datum, DatumRef, ScalarRefImpl, ToDatumRef};
+use risingwave_common::util::iter_util::ZipEqFast;
+use risingwave_pb::expr::expr_node::{RexNode, Type};
+use risingwave_pb::expr::ExprNode;
+
+use crate::expr::{build_from_prost, BoxedExpression, Expression};
+use crate::{bail, ensure, ExprError, Result};
+
+/// Returns a new array removing all the duplicates from the input array
+///
+/// ```sql
+/// array_distinct ( array anyarray) → array
+/// ```
+///
+/// Examples:
+///
+/// ```slt
+/// query T
+/// select array_distinct(array[NULL]);
+/// ----
+/// {NULL}
+///
+/// query T
+/// select array_distinct(array[1,2,1,1]);
+/// ----
+/// {1,2}
+///
+/// query T
+/// select array_distinct(array[1,2,1,NULL]);
+/// ----
+/// {1,2,NULL}
+/// query error polymorphic type
```

Comment:
```suggestion
/// {1,2,NULL}
///
/// query error polymorphic type
```
Need an empty line between 2 test cases.

Also for the `array_distinct(null::int[])` case, we need to write an explicit `NULL` as follows:
```
/// query T
/// select array_distinct(null::int[]);
/// ----
/// NULL
```

Reply:
Can you help me to implement this, for this, I have to check explicitly if input type is null, by using inputs[0].is_null() and then return a type or Ok(none), and then while evaluating, for matching arm None in evaluate function, I have to return a value which displays Null (this part I am having trouble with, what to return?)

Reply:
The current handling logic is already correct. There is no need to use `inputs[0].is_null()` in frontend, and the backend is already handling `null` with https://github.com/risingwavelabs/risingwave/pull/8315#discussion_r1123982833

The test failures you are getting is simply due to wrong formats of the test itself.
* When the input is `null::int[]`, the expected output is also `null::int[]`, which should be written as `NULL` in sqllogictest (slt), but [displays as nothing](https://www.postgresql.org/docs/15/app-psql.html#:~:text=The%20default%20is%20to%20print%20nothing%2C%20which%20can%20easily%20be%20mistaken%20for%20an%20empty%20string.) in `psql`. This may be where the confusion comes from:
```
dev=> select array_distinct(null::int[]);
 array_distinct 
----------------
 
(1 row)

dev=> \pset null 'NULL'
Null display is "NULL".
dev=> select array_distinct(null::int[]);
 array_distinct 
----------------
 NULL
(1 row)
```
* When the input is `null`, the expected error is already being reported in `psql`. We just need an empty line before the test case so that the slt test runner does not mistakenly think it is part of output of previous test case.
```
dev=> select array_distinct(null);
ERROR:  QueryError: Bind error: could not determine polymorphic type because input has type unknown
```

Reply:
> For input array_distinct(null::int[]), it is coming in `None` arm, what should I return instead, to display output as NULL, instead of empty value
```
Caused by:
  | query result mismatch:
  | [SQL] select array_distinct(null::int[]);
  | [Diff] (-expected\|+actual)
  | +   NULL
  | at e2e_test/generated/docslt/risingwave_expr/expr__expr__expr_array_distinct.slt:17
```
The returned `None` is already displayed as NULL in slt and as empty value in `psql`. But the test was expecting nothing:
```
/// query T
/// select array_distinct(null::int[]);
/// ----
```
And the fix is to update test expectation:
```
/// query T
/// select array_distinct(null::int[]);
/// ----
/// NULL
```